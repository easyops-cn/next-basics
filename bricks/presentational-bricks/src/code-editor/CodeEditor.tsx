import React, { useState, useEffect } from "react";
import AceEditor from "react-ace";
import "./brace/mode";
import "./brace/theme";
import style from "./index.module.css";
import classNames from "classnames";
import { find } from "lodash";

interface CodeEditorProps {
  theme: string;
  mode: string;
  setOptions?: Record<string, any>;
  editorProps?: Record<string, any>;
  editorStyle?: Record<string, any>;
  dataSource: any;
  configProps?: any;
  handleChange?: (value: any) => void;
  handleErrorChange?: (value: boolean) => void;
  handleBlur?: (value: any) => void;
  handleValidate?: (value: any) => void;
  required: boolean;
}

export function CodeEditor({
  theme,
  mode,
  setOptions,
  editorProps,
  dataSource,
  configProps,
  handleChange,
  handleBlur,
  handleErrorChange,
  required,
  editorStyle
}: CodeEditorProps): React.ReactElement {
  const [value, setValue] = useState(dataSource || "");
  const [hasError, setHasError] = useState(false);

  const handleOnChange = (newValue: string): void => {
    setValue(newValue);
  };

  useEffect(() => {
    handleChange(value);
  }, [value]);

  useEffect(() => {
    handleOnChange(dataSource || "");
  }, [dataSource]);

  const handleOnBlur = (): void => {
    handleBlur(value);
  };

  const onValidate = err => {
    let requiredError = false;
    if (required) {
      requiredError = !value;
    }
    setHasError(!!find(err, ["type", "error"]) || requiredError);
  };

  useEffect(() => {
    handleErrorChange(hasError);
  }, [hasError]);

  return (
    <AceEditor
      name="commands-editor"
      style={editorStyle}
      theme={theme}
      mode={mode}
      value={value}
      setOptions={setOptions}
      {...configProps}
      onChange={handleOnChange}
      onValidate={onValidate}
      onBlur={handleOnBlur}
      // Tips: Automatically scrolling cursor into view after selection change this will be disabled in the next version set editor.$blockScrolling = Infinity to disable this message
      editorProps={{ ...editorProps, $blockScrolling: Infinity }}
      className={classNames(style.aceContainer, {
        [style.hasError]: hasError
      })}
    />
  );
}
