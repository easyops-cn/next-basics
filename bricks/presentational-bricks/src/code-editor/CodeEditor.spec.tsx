import React from "react";
import { shallow } from "enzyme";
import { CodeEditor } from "./CodeEditor";
import AceEditor from "react-ace";

describe("CodeEditor", () => {
  it("should work", () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const handleErrorChange = jest.fn();
    const wrapper = shallow(
      <CodeEditor
        theme="tomorrow"
        mode="text"
        dataSource="123"
        editorStyle={{ width: "100%" }}
        setOptions={{ tabSize: 2 }}
        handleChange={onChange}
        handleBlur={onBlur}
        handleErrorChange={handleErrorChange}
        required={true}
      />
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ dataSource: "" });
    wrapper.find(AceEditor).invoke("onChange")("456");
    expect(wrapper.find(AceEditor).prop("value")).toBe("456");

    wrapper.find(AceEditor).invoke("onBlur")();
    expect(onBlur).toHaveBeenCalled();

    wrapper.find(AceEditor).invoke("onValidate")([]);
    expect(wrapper.find(AceEditor).prop("className")).not.toContain("hasError");

    wrapper.find(AceEditor).invoke("onValidate")([
      { type: "error", row: 1, column: 1, text: "err" }
    ]);
    expect(wrapper.find(AceEditor).prop("className")).toContain("hasError");

    wrapper.find(AceEditor).invoke("onChange")("");
    wrapper.find(AceEditor).invoke("onValidate")([]);
    expect(wrapper.find(AceEditor).prop("className")).toContain("hasError");
  });
});
