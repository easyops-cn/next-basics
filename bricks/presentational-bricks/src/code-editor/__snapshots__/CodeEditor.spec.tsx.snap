// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeEditor should work 1`] = `
<ForwardRef(ReactAce)
  className="aceContainer"
  editorProps={
    Object {
      "$blockScrolling": Infinity,
    }
  }
  mode="text"
  name="commands-editor"
  onBlur={[Function]}
  onChange={[Function]}
  onValidate={[Function]}
  setOptions={
    Object {
      "tabSize": 2,
    }
  }
  style={
    Object {
      "width": "100%",
    }
  }
  theme="tomorrow"
  value="123"
/>
`;
